docs/conftest.py

```diff
@@ -0,0 +1,52 @@
+#
+#  Copyright (C) 2022
+#  MIT
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+'''Pytest fixtures for running doctests
+
+Pytest looks for conftest.py files that define fixtures only in the
+directory tree that is being tested. When we run doctests in the
+documentation only, it will therefore not find that fixtures that are
+defined in sherpa/conftest.py, thus this file copies over the fixtures we
+need.
+'''
+import pytest
+from sherpa.utils.testing import get_datadir
+
+@pytest.fixture(autouse=True)
+def add_sherpa_test_data_dir(doctest_namespace):
```

Comment:
We have some documentation that calls routines that write to files - e.g. `sherpa.io.write_arrays`. Are these just going to be ones we skip, or can we handle this with some magic that runs the test in a temporary directory?

Reply:
[I'm basically running `pytest --doctest-plus sherpa/random/file.py` and looking at issues !]

Reply:
I have not spend too much thought on that, because I've not chosen one of those files as an example to test here. However, I'm not aware of magic that will do things just magically by just writing a few lines in this file.

The easiest is to just write to the current working directory. That should not be a problem in RTD or GH CI, because each build/test is done in a new, clean container. However, if you run the tests locally you would end up with artifacts lying around (unless you you tox and friends for local testing, which I don't).

In rst files, we can use the following directives (from https://github.com/astropy/pytest-doctestplus)
```
.. testsetup::

    >>> x = 42
```
and
```
.. testcleanup::

    >>> del x
```

They could be used to set the current working dir to a temporary directory and clean it up afterwards. Since each rst file is run as one test, `testsetup` would be inserted at the top and `testcleanup` at the bottom. Sphinx does not render any visible output for these directives. (Files may still end up laying around if the test fails in the middle and never gets to the clean up, but then that file needs to be looked at by hand anyway.)

In docstrings for individual functions that is another matter, we don't want our users to be bothered by sphinx directives. I think pytest setup and teardown (https://docs.pytest.org/en/6.2.x/xunit_setup.html) should work, but I have not tested that. If it's just one line that write a file and file is not needed further down, I would either skip that line (` # doctest: +SKIP`) or the entire doctest (see "skipping doctests" in https://github.com/astropy/pytest-doctestplus). 

While there are great advantages to running the doc examples with doctests and making sure they work, we don't have to bend over backwards if it's just not a good fit; the examples in the docs are examples first and tests only as far as it's useful.

Reply:
I'm just trying to understand the const-benefit ratio of this (ie how far do we go). Ta.